### ISSUE TITLE
Bad performance: Promise not fulfilled

### ISSUE THREAD

**Issue by nedark** [robertdavidgraham/masscan#444]
I'm trying to scan the internet for one port as fast as possible.

In the doc:
> This increases the rate to 100,000 packets/second, which will scan the entire Internet (minus excludes) in about 10 hours per port (or 655,360 hours if scanning all ports).

I tried with this value x10 (1,000,000 packets/second) and the estimated remaining time was about 11 hours. So I tried multiplying per 10 again (to 10,000,000 packets/second), and the estimated time didn't change.

Worse, after a few hours, the speed decreased, now this is the status information:
> rate: 23.27-kpps,  9.77% done,  43:17:54 remaining, found=12029

I am running in AWS EC2 instance. I feel unhappy of only using 23kbps of my 100mbps connection (0.022%).

Wasn't this supposed to be high performance?

**Comment by matthiasnehls** [comment]
The bad network @amazon 
Go rent a real box ;-)

**Comment by ravenium** [comment]
You're going to absolutely murder AWS's state tables - remember all EC2 instances are basically a 1:1 NAT.  I've generally found that masscan tends to start dropping ports around 10k pps in AWS, which isn't really masscan's fault but a consequence of the networking model.

Find a VPS with a public IP (and don't forget you need PF_RING above a certain PPS).

**Comment by XNerv** [comment]
Is possible to run in AWS with good results? What ype of instance can i use?

**Comment by ravenium** [comment]
Oh wow, it's been a while. I was able to get this to work! Basically, you can get about 100k-500k pps easily on a t3a.micro.  The trick is having a security group to allow 61000 (or whatever source port you use) inbound, then block it with iptables at the os level. That, together with a fixed source port in masscan, will get you the results and reliability you need.

Keep in mind your biggest bill will be vpc traffic.

**Comment by ravenium** [comment]
Apologies! I think my math is faulty.

I just did a cost breakdown for our AWS account for last month and the biggest cost is actually EC2 instances - we run 60-80 t3a.micros in an autoscale group that probably stands up ~6 hours a day.

In doing the re-mathing (ha!), I realized a few things:

- EC2 is about 3/4 of my cost but could be reduced (reserved instances, making use of burst capacity, spot pricing, etc).
- The previous "EC2 Other" cost I found is actually a few things - EBS, NAT gateways, etc which are likely contributing to that line item far more so than bandwidth. 
- Actual bandwidth costs are probably a drop in the bucket of that.  Amazon looks like they give you outbound at $0 for the first gig, then $0.09 per gig for up to 10 TB.

If you wanted to guess, try doing 72 byte SYN packets * 65k ports per host * number of IPs, then fudge a bit for additional payload interrogation.  I'm guessing it'll still be less than a few gigs for any sort of operation.

In short, you'll probably be ok - just keep an eye on your costs using AWS Billing and make sure you know what you're spending what on (which is the main battle of using AWS in any event).

Beyond that and the Security Group/Firewall rules, just make sure your activities fall into Amazon's "Penetration Test" authorized activities.  Make sure your pps rates don't affect whatever you're scanning, and that you have permission, etc.

**Comment by mzpqnxow** [comment]
If you really want to cover more than a few thousand hosts, get a dedicated bare-metal server. It's worth it

This solves not only any potential hardware constraints/challenges, but it solves the much more important network issue common to practically all VPS/"cloud" platforms

See the comment above about state table exhaustion. It's a very real problem as virtual hosting providers tend to track state *at least* because it's required for them to provide stateful firewalling/filtering for IaaS instances. Theoretically you could get around this with a 1:1 static port NAT and '--adapter-port' but that all depends on the hosting platform. By the time you find and test each, you might as well have just solved the problem properly with a dedicated leased or colocated bare-metal server

You have two options if you want to properly solve these problems (I recommend the first, it's less hassle)

1. leased: a server with one CPU, this will cost you from $75-$250/month
2. colocated: you'll have to ship your own hardware and pay for the rack space and bandwidth

These approximate prices assume modest hardware (2U with 1xCPU, 4-8 cores, 16GB RAM, 100GB storage- more than enough) and a 1Gbps connection, billed at [95th percentile](https://en.m.wikipedia.org/wiki/Burstable_billing) which is more than enough for scanning all of IPv4 on a single port quickly, even if done somewhat frequently

For another $100/month (depending on the provider) you can get a 10Gbps connection but that will be overkill

You can try and make Linode, DigitalOcean, EC2, Azure, etc. work and it may be fine for small sampling. In my experience that's a waste of time. If you want accurate results on larger samples, don't bother with that junk. Just pay up the $100-$200/month. If you're broke, share the cost with some friends or colleagues

tl;dr; expect to pay $100-$200month if you really want to do Internet-scale research and get accurate and reliable results. *Just because masscan says it's transmitting N pps doesn't mean the packets aren't getting dropped before they egress to the WAN*; and they often will be dropped if you're using a VPS/"cloud" platform, even with a dedicated IaaS instance
